Source/Core/VideoCommon/CPMemory.h

```diff
@@ -50,176 +52,239 @@ enum
-    u64 Tex0Coord : 2;
-    u64 Tex1Coord : 2;
-    u64 Tex2Coord : 2;
-    u64 Tex3Coord : 2;
-    u64 Tex4Coord : 2;
-    u64 Tex5Coord : 2;
-    u64 Tex6Coord : 2;
-    u64 Tex7Coord : 2;
-    u64 : 31;
+  RGB = 0,
+  RGBA = 1,
+};
+template <>
+struct fmt::formatter<ColorComponentCount> : EnumFormatter<ColorComponentCount::RGBA>
+{
+  formatter() : EnumFormatter({"3 (r, g, b)", "4 (r, g, b, a)"}) {}
+};
+
+enum class ColorFormat
+{
+  RGB565 = 0,    // 16b
+  RGB888 = 1,    // 24b
+  RGB888x = 2,   // 32b
+  RGBA4444 = 3,  // 16b
+  RGBA6666 = 4,  // 24b
+  RGBA8888 = 5,  // 32b
+};
+template <>
+struct fmt::formatter<ColorFormat> : EnumFormatter<ColorFormat::RGBA8888>
+{
+  static constexpr array_type names = {
+      "RGB 16 bits 565",   "RGB 24 bits 888",   "RGB 32 bits 888x",
+      "RGBA 16 bits 4444", "RGBA 24 bits 6666", "RGBA 32 bits 8888",
   };
+  formatter() : EnumFormatter(names) {}
+};
+
+enum class TexComponentCount
+{
+  S = 0,
+  ST = 1,
+};
+template <>
+struct fmt::formatter<TexComponentCount> : EnumFormatter<TexComponentCount::ST>
+{
+  formatter() : EnumFormatter({"1 (s)", "2 (s, t)"}) {}
+};
 
-  struct
+struct TVtxDesc
+{
+  union Low
+  {
+    // false: not present
+    // true: present
+    BitField<0, 1, bool, u32> PosMatIdx;
+    BitField<1, 1, bool, u32> Tex0MatIdx;
+    BitField<2, 1, bool, u32> Tex1MatIdx;
+    BitField<3, 1, bool, u32> Tex2MatIdx;
+    BitField<4, 1, bool, u32> Tex3MatIdx;
+    BitField<5, 1, bool, u32> Tex4MatIdx;
+    BitField<6, 1, bool, u32> Tex5MatIdx;
+    BitField<7, 1, bool, u32> Tex6MatIdx;
+    BitField<8, 1, bool, u32> Tex7MatIdx;
+    BitFieldArray<1, 1, 8, bool, u32> TexMatIdx;
+
+    BitField<9, 2, VertexComponentFormat> Position;
+    BitField<11, 2, VertexComponentFormat> Normal;
+    BitField<13, 2, VertexComponentFormat> Color0;
+    BitField<15, 2, VertexComponentFormat> Color1;
+    BitFieldArray<13, 2, 2, VertexComponentFormat> Color;
+
+    u32 Hex;
+  };
+  union High
   {
-    u32 Hex0, Hex1;
+    BitField<0, 2, VertexComponentFormat> Tex0Coord;
+    BitField<2, 2, VertexComponentFormat> Tex1Coord;
+    BitField<4, 2, VertexComponentFormat> Tex2Coord;
+    BitField<6, 2, VertexComponentFormat> Tex3Coord;
+    BitField<8, 2, VertexComponentFormat> Tex4Coord;
+    BitField<10, 2, VertexComponentFormat> Tex5Coord;
+    BitField<12, 2, VertexComponentFormat> Tex6Coord;
+    BitField<14, 2, VertexComponentFormat> Tex7Coord;
+    BitFieldArray<0, 2, 8, VertexComponentFormat> TexCoord;
+
+    u32 Hex;
   };
 
-  // Easily index into the Position..Tex7Coord fields.
-  u32 GetVertexArrayStatus(int idx) { return (Hex >> (9 + idx * 2)) & 0x3; }
+  Low low;
+  High high;
+
+  // This structure was originally packed into bits 0..32, using 33 total bits.
+  // The actual format has 17 bits in the low one and 16 bits in the high one,
+  // but the old format is still supported for compatibility.
+  u64 GetLegacyHex() const { return (low.Hex & 0x1FFFF) | (high.Hex << 17); }
```

Comment:
This is implemented wrong, as it does 32-bit math and then promotes to u64 afterwards.  `high.Hex` should be `u64(high.Hex)`.